shopping-cart/shopping-cart-scala/shopping-cart/src/main/scala/com/example/shoppingcart/impl/ShoppingCart.scala

```diff
@@ -0,0 +1,316 @@
+package com.example.shoppingcart.impl
+
+import java.time.Instant
+
+import akka.Done
+import akka.actor.typed.{ActorRef, Behavior}
+import akka.cluster.sharding.typed.scaladsl._
+import akka.persistence.typed.PersistenceId
+import akka.persistence.typed.scaladsl.RetentionCriteria
+import akka.persistence.typed.scaladsl.{Effect, EventSourcedBehavior, ReplyEffect}
+import com.lightbend.lagom.scaladsl.persistence.{AggregateEvent, AggregateEventTag }
+import com.lightbend.lagom.scaladsl.playjson.{JsonSerializer, JsonSerializerRegistry}
+import play.api.libs.json.{Format, _}
+
+import scala.collection.immutable.Seq
+import com.lightbend.lagom.scaladsl.persistence.AkkaTaggerAdapter
+
+final case class ShoppingCart(items: Map[String, Int], checkedOut: Boolean) {
+
+  //The shopping cart behavior changes if it's checked out or not. The command handles are different for each case.
+  def applyCommand(cmd: ShoppingCartCommand): ReplyEffect[ShoppingCartEvent, ShoppingCart] = 
+    if (checkedOut) applyCommandForCheckedOut(cmd)
+    else applyCommandForOpen(cmd)
+
+  // command handlers for open shopping cart
+  private def applyCommandForOpen(cmd: ShoppingCartCommand): ReplyEffect[ShoppingCartEvent, ShoppingCart] = 
+    cmd match {
+      case AddItem(itemId, quantity, replyTo) =>
+        if (items.contains(itemId))
+          Effect.reply(replyTo)(Rejected(s"Item '$itemId' was already added to this shopping cart"))
+        else if (quantity <= 0)
+          Effect.reply(replyTo)(Rejected("Quantity must be greater than zero"))
+        else
+          Effect
+            .persist(ItemAdded(itemId, quantity))
+            .thenReply(replyTo)(updatedCart => Accepted(toSummary(updatedCart)))
+
+      case RemoveItem(itemId, replyTo) =>
+        if (items.contains(itemId))
+          Effect
+            .persist(ItemRemoved(itemId))
+            .thenReply(replyTo)(updatedCart => Accepted(toSummary(updatedCart)))
+        else
+          Effect.reply(replyTo)(Accepted(toSummary(this))) // removing an item is idempotent
+
+      case AdjustItemQuantity(itemId, quantity, replyTo) =>
+        if(items.contains(itemId))
+          Effect
+            .persist(ItemQuantityAdjusted(itemId, quantity))
+            .thenReply(replyTo)(updatedCart => Accepted(toSummary(updatedCart)))
+        else
+          Effect.reply(replyTo)(Rejected(s"Cannot adjust quantity for item '$itemId'. Item not present on cart"))
```

Comment:
(nitpick)
If the `amount` of `item`s in the cart is the same than the value on the command then, no events should be persisted and an `Accepted` be returned.